src/Eppie.App/Eppie.AI/Utils/GenAIModel.cs

```diff
@@ -0,0 +1,305 @@
+
+                    if (string.IsNullOrWhiteSpace(_template.User))
+                    {
+                        prompt.Append(msgText);
+                    }
+                    else
+                    {
+                        prompt.Append(_template.User.Replace(TEMPLATE_PLACEHOLDER, msgText));
+                    }
+                }
+                else if (message.Role == ChatRole.Assistant)
+                {
+                    if (string.IsNullOrWhiteSpace(_template.Assistant))
+                    {
+                        prompt.Append(message.Text);
+                    }
+                    else
+                    {
+                        prompt.Append(_template.Assistant.Replace(TEMPLATE_PLACEHOLDER, message.Text));
+                    }
+                }
+            }
+
+            if (!string.IsNullOrWhiteSpace(_template.Assistant))
+            {
+                var substringIndex = _template.Assistant.IndexOf(TEMPLATE_PLACEHOLDER, StringComparison.InvariantCulture);
+                prompt.Append(_template.Assistant.Substring(0, substringIndex));
+            }
+
+            return prompt.ToString();
+        }
+
+        public async Task<ChatCompletion> CompleteAsync(IList<ChatMessage> chatMessages, ChatOptions options = null, CancellationToken ct = default)
+        {
+            var result = string.Empty;
+            await foreach (var part in CompleteStreamingAsync(chatMessages, options, ct).ConfigureAwait(false))
+            {
+                result += part.Text;
+            }
+
+            return new ChatCompletion(new ChatMessage(ChatRole.Assistant, result));
+        }
+
+        public async IAsyncEnumerable<StreamingChatCompletionUpdate> CompleteStreamingAsync(IList<ChatMessage> chatMessages, ChatOptions options = null, [EnumeratorCancellation] CancellationToken ct = default)
+        {
+            var prompt = GetPrompt(chatMessages);
+
+            if (!IsReady)
+            {
+                throw new InvalidOperationException("Model is not ready");
+            }
+
+            using var generatorParams = new GeneratorParams(_model);
+
+            using var sequences = _tokenizer.Encode(prompt);
+
+            void TransferMetadataValue(string propertyName, object defaultValue)
+            {
+                object val = null;
+                options?.AdditionalProperties?.TryGetValue(propertyName, out val);
+
+                val ??= defaultValue;
+
+                if (val is int intVal)
+                {
+                    generatorParams.SetSearchOption(propertyName, intVal);
+                }
+                else if (val is float floatVal)
+                {
+                    generatorParams.SetSearchOption(propertyName, floatVal);
+                }
+                else if (val is bool boolVal)
+                {
+                    generatorParams.SetSearchOption(propertyName, boolVal);
+                }
+            }
+
+            if (options != null)
```

Comment:
Is this `if` needed here?

Reply:
This file from AI-Dev-Gallery, as soon as they make a library we can plug in, we will get rid of this copy.